library/cpp/include/wavemap/core/utils/profile/resource_monitor.h

```diff
@@ -0,0 +1,177 @@
+/**
+ * @brief Monitors system resource usage, including CPU time, wall time,
+ *        and RAM usage.
+ *
+ * The `ResourceMonitor` class tracks CPU and wall clock time over timed
+ * episodes, much like wavemap's Stopwatch class. It also provides
+ * functionality to retrieve the total RAM usage of the current process.
+ */
+class ResourceMonitor {
+ public:
+  /**
+   * @brief Starts a new CPU and wall time monitoring episode.
+   *
+   * Records the CPU and wall clock start times for the current episode.
+   * If monitoring is already running, calling `start()` has no effect.
+   */
+  void start();
+
+  /**
+   * @brief Stops timing the current episode.
+   *
+   * Records the end CPU and wall clock times for the current episode, updating
+   * the last episode duration and total accumulated duration. If no episode is
+   * in progress, calling `stop()` has no effect.
+   */
+  void stop();
+
+  /**
+   * @brief Checks if the stopwatch is currently running.
+   *
+   * @return `true` if the stopwatch is running, `false` otherwise.
+   */
+  bool isRunning() const { return running_; }
+
+  /**
+   * @brief Gets the current RAM usage of the application.
+   *
+   * @return The current RAM usage in kilobytes, or `std::nullopt` (an empty
+   *         optional) if retrieving RAM usage is not supported on the given
+   *         platform.
+   */
+  static std::optional<size_t> getCurrentRamUsageInKB();
+
+  /**
+   * @brief Gets the CPU time duration of the last episode.
+   *
+   * @return The CPU time duration of the last episode in seconds.
+   *
+   * The value represents the CPU time elapsed during the most recently
+   * completed pair of `start()` and `stop()` calls. If no episode has been
+   * completed, this returns 0.
+   */
+  double getLastEpisodeCpuTime() const {
+    return time::to_seconds<double>(last_episode_cpu_duration_);
+  }
+
+  /**
+   * @brief Gets the wall clock time duration of the last episode.
+   *
+   * @return The wall clock time duration of the last episode in seconds.
+   *
+   * The value represents the real-world time elapsed during the most recently
+   * completed pair of `start()` and `stop()` calls. If no episode has been
+   * completed, this returns 0.
+   */
+  double getLastEpisodeWallTime() const {
+    return time::to_seconds<double>(last_episode_wall_duration_);
+  }
+
+  /**
+   * @brief Get the last episode's resource usage stats formatted as a string.
+   *
+   * @return A string with the CPU time, wall time, and RAM usage statistics.
+   *
+   * The returned string provides a human-readable summary of the resource
+   * usage for the most recently completed episode. CPU and wall times are
+   * displayed in seconds with two decimal places, while RAM usage is reported
+   * in kilobytes. If RAM usage information is unavailable, it will be labeled
+   * as "Unknown". Each statistic is printed on a new line, with a leading `*`.
+   */
+  std::string getLastEpisodeResourceUsageStats() const;
```

Comment:
@LionelOtt, this method and its `getTotalResourceUsageStats` sibling can now be used to format the stats as a string.